### PR TITLE
Add various enhancements

### PR DESCRIPTION
**Pull request by natlibfi-arlehiko**
**Tested only with Graylog 3**

Add support for:
- Extractors
- Graylog SSO auth plugin
- Users

Fix/enhance:
- Index sets
- Roles
- Streams

### CODE REVIEW

**Comment by philomory** [comment]
Thanks, I'll review and merge this when I get back to my office on the 12th

**Review by philomory — CHANGES_REQUESTED** [review]
Overall looks pretty good but I did have a few questions about the way certain things were handled.

File: lib/puppet/provider/graylog_api.rb [hunk 1]
```diff
@@ -174,12 +183,33 @@ def simple_flush(path,params)
...
+  def user_flush(params)
+    path = 'users'
+
+    if !defined?(@action)
+      params = params
+      .select { |k| ![:username, :password].include?(k) }          
+    end
+
+    simple_flush('users', params)
+  end
```
[the remark this answers is not in the document]
Comment (philomory) [reply lib/puppet/provider/graylog_api.rb:213]: Also, the field filtering can be much more simply written as `params -= [:username, :password] unless @action`.

File: lib/puppet/provider/graylog_index_set/graylog_api.rb [hunk 2]
```diff
@@ -46,7 +47,9 @@ def flush
       index_analyzer: resource[:index_analyzer],
       index_optimization_max_num_segments: resource[:max_segments],
       index_optimization_disabled: resource[:disable_index_optimization],
+      creation_date: '2000-01-01T00:00:00.00Z',
```
[the remark this answers is not in the document]
Comment (natlibfi-arlehiko) [reply lib/puppet/provider/graylog_index_set/graylog_api.rb]: 'twas some development time laziness :) Fixed now.

File: lib/puppet/provider/graylog_api.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (natlibfi-arlehiko) [reply lib/puppet/provider/graylog_api.rb]: > Also, the field filtering can be much more simply written as `params -= [:username, :password] unless @action`.

I got `undefined method `-' for nil:NilClass (NoMethodError)` for the `-=` operator. Maybe it's supported only in newer Ruby versions or something.

File: lib/puppet/provider/graylog_plugin_auth_sso/graylog_api.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+Puppet::Type.type(:graylog_plugin_auth_sso).provide(:graylog_api, parent: Puppet::Provider::GraylogAPI) do
+
+  mk_resource_methods
+
+  def self.instances
+      data = get('plugins/org.graylog.plugins.auth.sso/config')
+      [new(
+        ensure: :present,
+        name: 'org.graylog.plugins.auth.sso/config',
```
Comment (philomory) [inline lib/puppet/provider/graylog_plugin_auth_sso/graylog_api.rb:11]:
```suggestion
        name: 'sso',
```

File: lib/puppet/type/graylog_plugin_auth_sso.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+Puppet::Type.newtype(:graylog_plugin_auth_sso) do
+
+  desc <<-END_OF_DOC
+    @summary
+      SSO authentication plugin configuration
+
+    SSO authentication pluging configuration definition.
+
+    @example
+      graylog_auth_sso_plugin_config { 'org.graylog.plugins.foo/config':
```
Comment (philomory) [inline lib/puppet/type/graylog_plugin_auth_sso.rb:12]:
```suggestion
      graylog_auth_sso_plugin_config { 'sso':
```

**Review by philomory — CHANGES_REQUESTED** [review]
Just a few tiny changes left, thanks for your work on this!

**Comment by natlibfi-arlehiko** [comment]
Thanks! Here you go again :)

**Review by philomory — APPROVED** [review]
Everything looks good to me at this point! Thanks for your help!

**Comment by philomory** [comment]
I'll release a new version to puppet forge shortly.

**Comment by philomory** [comment]
Version 0.4.0 has been released

**Pull request merged**